File: data/combination/combinations.py
def combine(n: int, k: int) -> list[list[int]]:
    """
    Given two integers n and k, return all possible combinations of k numbers chosen from [1, n].
    Time Complexity: O(C(n, k))
    Space Complexity: O(k)
    """
    res = []
    def backtrack(start, comb):
        if len(comb) == k:
            res.append(comb[:])
            return
        for i in range(start, n + 1):
            comb.append(i)
            backtrack(i + 1, comb)
            comb.pop()

    backtrack(1, [])
    return res

File: data/combination/combinations.cpp
#include <vector>

using namespace std;

class SolutionCombinations {
    vector<vector<int>> res;
    void backtrack(int n, int k, int start, vector<int>& comb) {
        if (comb.size() == k) {
            res.push_back(comb);
            return;
        }
        for (int i = start; i <= n; ++i) {
            comb.push_back(i);
            backtrack(n, k, i + 1, comb);
            comb.pop_back();
        }
    }
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> comb;
        backtrack(n, k, 1, comb);
        return res;
    }
};
